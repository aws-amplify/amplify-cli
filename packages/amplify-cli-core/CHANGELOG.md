# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 1.1.0 (2020-07-23)


### Features

* feature flag implementation ([#4891](https://github.com/aws-amplify/amplify-cli/issues/4891)) ([4639450](https://github.com/aws-amplify/amplify-cli/commit/463945029cfe861f74986d9a8b9af6b827d2063d))
