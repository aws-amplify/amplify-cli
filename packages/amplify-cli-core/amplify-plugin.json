{
    "name": "core",
    "type": "core",
    "commands": [
        "plugin",
        "diagnose",
        "version"
    ],
    "subscriptions":[]
}