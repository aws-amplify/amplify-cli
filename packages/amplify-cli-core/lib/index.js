"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExeInfo = void 0;
const ExeInfo = __importStar(require("./exeInfo"));
exports.ExeInfo = ExeInfo;
__exportStar(require("./banner-message"), exports);
__exportStar(require("./category-interfaces"), exports);
__exportStar(require("./cfnUtilities"), exports);
__exportStar(require("./cliConstants"), exports);
__exportStar(require("./cliContext"), exports);
__exportStar(require("./cliContextEnvironmentProvider"), exports);
__exportStar(require("./cliEnvironmentProvider"), exports);
__exportStar(require("./cliGetCategories"), exports);
__exportStar(require("./cliRemoveResourcePrompt"), exports);
__exportStar(require("./cliViewAPI"), exports);
__exportStar(require("./customPoliciesUtils"), exports);
__exportStar(require("./deploymentSecretsHelper"), exports);
__exportStar(require("./deploymentState"), exports);
__exportStar(require("./errors"), exports);
__exportStar(require("./exitOnNextTick"), exports);
__exportStar(require("./feature-flags"), exports);
__exportStar(require("./hooks"), exports);
__exportStar(require("./isCI"), exports);
__exportStar(require("./isPackaged"), exports);
__exportStar(require("./jsonUtilities"), exports);
__exportStar(require("./overrides-manager"), exports);
__exportStar(require("./permissionsBoundaryState"), exports);
__exportStar(require("./plugin-facade"), exports);
__exportStar(require("./serviceSelection"), exports);
__exportStar(require("./spinner"), exports);
__exportStar(require("./state-manager"), exports);
__exportStar(require("./tags"), exports);
__exportStar(require("./types"), exports);
__exportStar(require("./utils"), exports);
__exportStar(require("./errors/amplify-error"), exports);
__exportStar(require("./errors/amplify-fault"), exports);
__exportStar(require("./errors/amplify-exception"), exports);
__exportStar(require("./errors/project-not-initialized-error"), exports);
__exportStar(require("./context/context-extensions"), exports);
__exportStar(require("./context/plugin-platform"), exports);
__exportStar(require("./context/amplify-event"), exports);
__exportStar(require("./context/plugin-collection"), exports);
__exportStar(require("./context/plugin-info"), exports);
__exportStar(require("./context/plugin-manifest"), exports);
__exportStar(require("./context/plugin-verification-result"), exports);
__exportStar(require("./constants"), exports);
__exportStar(require("./help"), exports);
__exportStar(require("./amplify-node-pkg-detector"), exports);
//# sourceMappingURL=index.js.map