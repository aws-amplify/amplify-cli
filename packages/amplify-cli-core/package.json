{
  "name": "amplify-cli-core",
  "version": "1.24.1",
  "description": "Amplify CLI Core",
  "repository": {
    "type": "git",
    "url": "https://github.com/aws-amplify/amplify-cli.git",
    "directory": "packages/amplify-cli-core"
  },
  "author": "Amazon Web Services",
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "keywords": [
    "graphql",
    "appsync",
    "aws"
  ],
  "scripts": {
    "test": "jest",
    "test-ci": "jest --ci -i",
    "build": "tsc",
    "watch": "tsc -w",
    "clean": "rimraf lib tsconfig.tsbuildinfo"
  },
  "dependencies": {
    "ajv": "^6.12.3",
    "amplify-cli-logger": "1.1.0",
    "ci-info": "^2.0.0",
    "cloudform-types": "^4.2.0",
    "dotenv": "^8.2.0",
    "folder-hash": "^4.0.1",
    "fs-extra": "^8.1.0",
    "globby": "^11.0.3",
    "hjson": "^3.2.1",
    "js-yaml": "^4.0.0",
    "lodash": "^4.17.21",
    "node-fetch": "^2.6.1",
    "open": "^7.3.1",
    "proxy-agent": "^4.0.1",
    "semver": "^7.3.5",
    "which": "^2.0.2"
  },
  "devDependencies": {
    "@types/folder-hash": "^4.0.0",
    "@types/fs-extra": "^8.0.1",
    "@types/hjson": "^2.4.2",
    "@types/json-schema": "^7.0.5",
    "@types/lodash": "^4.14.149",
    "@types/node": "^12.12.6",
    "@types/rimraf": "^3.0.0",
    "@types/uuid": "^8.0.0",
    "amplify-function-plugin-interface": "1.9.0",
    "nock": "^13.0.11",
    "rimraf": "^3.0.0"
  },
  "jest": {
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "testRegex": "(src/__tests__/.*.test.ts)$",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "jsx",
      "json",
      "node"
    ],
    "collectCoverage": true
  }
}
