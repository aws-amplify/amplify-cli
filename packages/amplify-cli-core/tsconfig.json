{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "outDir": "./lib",
    "rootDir": "./src",
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
  }
}
