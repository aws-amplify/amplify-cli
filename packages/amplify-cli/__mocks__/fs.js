const fs = jest.genMockFromModule('fs');

function readdirSync() {
  return ['file1', 'file2'];
}

function unlinkSync() {
  return null;
}

function readFileSync() {
  return {};
}

function statSync() {
  return { isDirectory: () => {} };
}

fs.readdirSync = readdirSync;
fs.readFileSync = readFileSync;
fs.statSync = statSync;
fs.unlinkSync = unlinkSync;

module.exports = fs;
