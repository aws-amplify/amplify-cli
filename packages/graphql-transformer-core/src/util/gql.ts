import { parse } from 'graphql';
export function gql(literals: TemplateStringsArray, ...placeholders: string[]) {
  const interleaved = [];
  for (let i = 0; i < placeholders.length; i++) {
    interleaved.push(literals[i]);
    interleaved.push(placeholders[i]);
  }
  interleaved.push(literals[literals.length - 1]);
  const actualSchema = interleaved.join('');
  return parse(actualSchema);
}
