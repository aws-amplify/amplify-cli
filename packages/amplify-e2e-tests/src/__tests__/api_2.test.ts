import { JSONUtilities } from 'amplify-cli-core';
import {
  addApi,
  addApiWithBlankSchemaAndConflictDetection,
  addApiWithoutSchema,
  amplifyPush,
  amplifyPushUpdate,
  apiDisableDataStore,
  apiEnableDataStore,
  apiGqlCompile,
  createNewProjectDir,
  deleteProject,
  deleteProjectDir,
  enableAdminUI,
  getAppSyncApi,
  getLocalEnvInfo,
  getProjectMeta,
  getTransformConfig,
  initJSProjectWithProfile,
  updateApiSchema,
  updateAPIWithResolutionStrategyWithModels,
  setCustomRolesConfig,
} from 'amplify-e2e-core';
import AWSAppSyncClient, { AUTH_TYPE } from 'aws-appsync';
import { existsSync, readFileSync } from 'fs';
import gql from 'graphql-tag';
import { TRANSFORM_CURRENT_VERSION } from 'graphql-transformer-core';
import _ from 'lodash';
import * as path from 'path';
const providerName = 'awscloudformation';

// to deal with bug in cognito-identity-js
(global as any).fetch = require('node-fetch');
// to deal with subscriptions in node env
(global as any).WebSocket = require('ws');

describe('amplify add api (GraphQL)', () => {
  let projRoot: string;
  beforeEach(async () => {
    projRoot = await createNewProjectDir('graphql-api');
  });

  afterEach(async () => {
    const metaFilePath = path.join(projRoot, 'amplify', '#current-cloud-backend', 'amplify-meta.json');
    if (existsSync(metaFilePath)) {
      await deleteProject(projRoot);
    }
    deleteProjectDir(projRoot);
  });

  it('init a project with conflict detection enabled and a schema with @key, test update mutation', async () => {
    const name = `keyconflictdetection`;
    await initJSProjectWithProfile(projRoot, { name });
    await addApiWithBlankSchemaAndConflictDetection(projRoot, { transformerVersion: 1 });
    await updateApiSchema(projRoot, name, 'key-conflict-detection.graphql');
    await amplifyPush(projRoot);

    const meta = getProjectMeta(projRoot);
    const region = meta['providers'][providerName]['Region'] as string;
    const { output } = meta.api[name];
    const url = output.GraphQLAPIEndpointOutput as string;
    const apiKey = output.GraphQLAPIKeyOutput as string;

    const appSyncClient = new AWSAppSyncClient({
      url,
      region,
      disableOffline: true,
      auth: {
        type: AUTH_TYPE.API_KEY,
        apiKey,
      },
    });

    const createMutation = /* GraphQL */ `
      mutation CreateNote($input: CreateNoteInput!, $condition: ModelNoteConditionInput) {
        createNote(input: $input, condition: $condition) {
          noteId
          note
          _version
          _deleted
          _lastChangedAt
          createdAt
          updatedAt
        }
      }
    `;
    const createInput = {
      input: {
        noteId: '1',
        note: 'initial note',
      },
    };
    const createResult: any = await appSyncClient.mutate({
      mutation: gql(createMutation),
      fetchPolicy: 'no-cache',
      variables: createInput,
    });

    const updateMutation = /* GraphQL */ `
      mutation UpdateNote($input: UpdateNoteInput!, $condition: ModelNoteConditionInput) {
        updateNote(input: $input, condition: $condition) {
          noteId
          note
          _version
          _deleted
          _lastChangedAt
          createdAt
          updatedAt
        }
      }
    `;
    const createResultData = createResult.data as any;
    const updateInput = {
      input: {
        noteId: createResultData.createNote.noteId,
        note: 'note updated',
        _version: createResultData.createNote._version,
      },
    };

    const updateResult: any = await appSyncClient.mutate({
      mutation: gql(updateMutation),
      fetchPolicy: 'no-cache',
      variables: updateInput,
    });
    const updateResultData = updateResult.data as any;

    expect(updateResultData).toBeDefined();
    expect(updateResultData.updateNote).toBeDefined();
    expect(updateResultData.updateNote.noteId).toEqual(createResultData.createNote.noteId);
    expect(updateResultData.updateNote.note).not.toEqual(createResultData.createNote.note);
    expect(updateResultData.updateNote._version).not.toEqual(createResultData.createNote._version);
    expect(updateResultData.updateNote.note).toEqual(updateInput.input.note);
  });

  it('init a project with conflict detection enabled and toggle disable', async () => {
    const name = `conflictdetection`;
    await initJSProjectWithProfile(projRoot, { name });
    await addApiWithBlankSchemaAndConflictDetection(projRoot, { transformerVersion: 1 });
    await updateApiSchema(projRoot, name, 'simple_model.graphql');

    await amplifyPush(projRoot);

    const meta = getProjectMeta(projRoot);
    const { output } = meta.api[name];
    const { GraphQLAPIIdOutput, GraphQLAPIEndpointOutput, GraphQLAPIKeyOutput } = output;
    const { graphqlApi } = await getAppSyncApi(GraphQLAPIIdOutput, meta.providers.awscloudformation.Region);

    expect(GraphQLAPIIdOutput).toBeDefined();
    expect(GraphQLAPIEndpointOutput).toBeDefined();
    expect(GraphQLAPIKeyOutput).toBeDefined();

    expect(graphqlApi).toBeDefined();
    expect(graphqlApi.apiId).toEqual(GraphQLAPIIdOutput);

    const transformConfig = getTransformConfig(projRoot, name);
    expect(transformConfig).toBeDefined();
    expect(transformConfig.Version).toBeDefined();
    expect(transformConfig.Version).toEqual(TRANSFORM_CURRENT_VERSION);
    expect(transformConfig.ResolverConfig).toBeDefined();
    expect(transformConfig.ResolverConfig.project).toBeDefined();
    expect(transformConfig.ResolverConfig.project.ConflictDetection).toEqual('VERSION');
    expect(transformConfig.ResolverConfig.project.ConflictHandler).toEqual('AUTOMERGE');

    // remove datastore feature
    await apiDisableDataStore(projRoot, {});
    await amplifyPushUpdate(projRoot);
    const disableDSConfig = getTransformConfig(projRoot, name);
    expect(disableDSConfig).toBeDefined();
    expect(_.isEmpty(disableDSConfig.ResolverConfig)).toBe(true);
  });

  it('init a project with conflict detection enabled and admin UI enabled to generate datastore models in the cloud', async () => {
    const name = `dsadminui`;
    await initJSProjectWithProfile(projRoot, { disableAmplifyAppCreation: false, name });

    let meta = getProjectMeta(projRoot);
    const appId = meta.providers?.[providerName]?.AmplifyAppId;
    const region = meta.providers?.[providerName]?.Region;

    expect(appId).toBeDefined();

    const localEnvInfo = getLocalEnvInfo(projRoot);
    const envName = localEnvInfo.envName;

    // setupAdminUI
    await enableAdminUI(appId, envName, region);

    await addApiWithBlankSchemaAndConflictDetection(projRoot, { transformerVersion: 1 });
    await updateApiSchema(projRoot, name, 'simple_model.graphql');
    await amplifyPush(projRoot);

    meta = getProjectMeta(projRoot);

    const { output } = meta.api[name];
    const { GraphQLAPIIdOutput, GraphQLAPIEndpointOutput, GraphQLAPIKeyOutput } = output;

    expect(GraphQLAPIIdOutput).toBeDefined();
    expect(GraphQLAPIEndpointOutput).toBeDefined();
    expect(GraphQLAPIKeyOutput).toBeDefined();

    const { graphqlApi } = await getAppSyncApi(GraphQLAPIIdOutput, meta.providers.awscloudformation.Region);

    expect(graphqlApi).toBeDefined();
    expect(graphqlApi.apiId).toEqual(GraphQLAPIIdOutput);
  });

  it('init a sync enabled project and update conflict resolution strategy', async () => {
    const name = `syncenabled`;
    await initJSProjectWithProfile(projRoot, { name });
    await addApiWithBlankSchemaAndConflictDetection(projRoot, { transformerVersion: 1 });
    await updateApiSchema(projRoot, name, 'simple_model.graphql');

    let transformConfig = getTransformConfig(projRoot, name);
    expect(transformConfig).toBeDefined();
    expect(transformConfig.ResolverConfig).toBeDefined();
    expect(transformConfig.ResolverConfig.project).toBeDefined();
    expect(transformConfig.ResolverConfig.project.ConflictDetection).toEqual('VERSION');
    expect(transformConfig.ResolverConfig.project.ConflictHandler).toEqual('AUTOMERGE');

    await updateAPIWithResolutionStrategyWithModels(projRoot, {});

    transformConfig = getTransformConfig(projRoot, name);
    expect(transformConfig).toBeDefined();
    expect(transformConfig.Version).toBeDefined();
    expect(transformConfig.Version).toEqual(TRANSFORM_CURRENT_VERSION);
    expect(transformConfig.ResolverConfig).toBeDefined();
    expect(transformConfig.ResolverConfig.project).toBeDefined();
    expect(transformConfig.ResolverConfig.project.ConflictDetection).toEqual('VERSION');
    expect(transformConfig.ResolverConfig.project.ConflictHandler).toEqual('OPTIMISTIC_CONCURRENCY');

    await amplifyPush(projRoot);
    const meta = getProjectMeta(projRoot);
    const { output } = meta.api[name];
    const { GraphQLAPIIdOutput, GraphQLAPIEndpointOutput, GraphQLAPIKeyOutput } = output;
    const { graphqlApi } = await getAppSyncApi(GraphQLAPIIdOutput, meta.providers.awscloudformation.Region);

    expect(GraphQLAPIIdOutput).toBeDefined();
    expect(GraphQLAPIEndpointOutput).toBeDefined();
    expect(GraphQLAPIKeyOutput).toBeDefined();

    expect(graphqlApi).toBeDefined();
    expect(graphqlApi.apiId).toEqual(GraphQLAPIIdOutput);
  });

  it('init a datastore enabled project and then remove datastore config in update', async () => {
    const name = 'withoutdatastore';
    await initJSProjectWithProfile(projRoot, { name });
    await addApiWithoutSchema(projRoot, { transformerVersion: 1 });
    await updateApiSchema(projRoot, name, 'simple_model.graphql');
    await amplifyPush(projRoot);

    const meta = getProjectMeta(projRoot);
    const { output } = meta.api[name];
    const { GraphQLAPIIdOutput, GraphQLAPIEndpointOutput, GraphQLAPIKeyOutput } = output;
    const { graphqlApi } = await getAppSyncApi(GraphQLAPIIdOutput, meta.providers.awscloudformation.Region);

    expect(GraphQLAPIIdOutput).toBeDefined();
    expect(GraphQLAPIEndpointOutput).toBeDefined();
    expect(GraphQLAPIKeyOutput).toBeDefined();

    expect(graphqlApi).toBeDefined();
    expect(graphqlApi.apiId).toEqual(GraphQLAPIIdOutput);

    // check project doesn't have datastore
    const withoutDSConfig = getTransformConfig(projRoot, name);
    expect(withoutDSConfig).toBeDefined();
    expect(_.isEmpty(withoutDSConfig.ResolverConfig)).toBe(true);

    // amplify update api to enable datastore
    await apiEnableDataStore(projRoot, {});
    let transformConfigWithDS = getTransformConfig(projRoot, name);
    expect(transformConfigWithDS).toBeDefined();
    expect(transformConfigWithDS.ResolverConfig).toBeDefined();
    expect(transformConfigWithDS.ResolverConfig.project).toBeDefined();
    expect(transformConfigWithDS.ResolverConfig.project.ConflictHandler).toEqual('AUTOMERGE');
    expect(transformConfigWithDS.ResolverConfig.project.ConflictDetection).toEqual('VERSION');
  });

  it('init a project and add custom iam roles - local test with gql v2', async () => {
    const name = 'customadminroles';
    await initJSProjectWithProfile(projRoot, { name });
    await addApi(projRoot, { transformerVersion: 2, IAM: {}, 'Amazon Cognito User Pool': {} });
    updateApiSchema(projRoot, name, 'cognito_simple_model.graphql');
    await apiGqlCompile(projRoot);
    const createResolver = path.join(
      projRoot,
      'amplify',
      'backend',
      'api',
      name,
      'build',
      'resolvers',
      'Mutation.createTodo.auth.1.req.vtl',
    );
    const beforeAdminConfig = readFileSync(createResolver).toString();
    expect(beforeAdminConfig).toMatchSnapshot();

    const customRolesConfig = {
      adminRoleNames: ['myAdminRoleName'],
    };
    setCustomRolesConfig(projRoot, name, customRolesConfig);
    await apiGqlCompile(projRoot);
    const afterAdminConfig = readFileSync(createResolver).toString();
    expect(afterAdminConfig).toMatchSnapshot();
    expect(beforeAdminConfig).not.toEqual(afterAdminConfig);
  });

  // TODO: Disabling for now until further conversation.
  // it('inits a project with a simple model with deletion protection enabled and then migrates the api', async () => {
  //   const projectName = 'retaintables';
  //   const initialSchema = 'simple_model.graphql';
  //   console.log(projRoot);
  //   await initJSProjectWithProfile(projRoot, { name: projectName });
  //   await addApiWithSchema(projRoot, initialSchema);
  //   updateConfig(projRoot, projectName, {
  //     TransformerOptions: {
  //       '@model': { EnableDeletionProtection: true }
  //     }
  //   });
  //   await amplifyPush(projRoot);
  //   const projectMeta = getProjectMeta(projRoot);
  //   const region = projectMeta.providers.awscloudformation.Region;
  //   const { output } = projectMeta.api[projectName];
  //   const { GraphQLAPIIdOutput, GraphQLAPIEndpointOutput, GraphQLAPIKeyOutput } = output;
  //   await expect(GraphQLAPIIdOutput).toBeDefined()
  //   await expect(GraphQLAPIEndpointOutput).toBeDefined()
  //   await expect(GraphQLAPIKeyOutput).toBeDefined()
  //   await deleteProject(projRoot);
  //   const tableName = `Todo-${GraphQLAPIIdOutput}-integtest`;
  //   const table = await getTable(tableName, region);
  //   expect(table.Table).toBeDefined()
  //   if (table.Table) {
  //     const del = await deleteTable(tableName, region);
  //     expect(del.TableDescription).toBeDefined()
  //   }
  // });
});
