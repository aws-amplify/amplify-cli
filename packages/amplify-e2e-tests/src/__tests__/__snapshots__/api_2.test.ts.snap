// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amplify add api (GraphQL) init a project and add custom iam roles - local test with gql v2 1`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $inputFields = $util.parseJson($util.toJson($ctx.args.input.keySet())) )
#set( $isAuthorized = false )
#set( $allowedFields = [] )
#if( $util.authType() == \\"IAM Authorization\\" )
$util.unauthorized()
#end
#if( $util.authType() == \\"User Pool Authorization\\" )
  #if( !$isAuthorized )
    #set( $ownerEntity0 = $util.defaultIfNull($ctx.args.input.owner, null) )
    #set( $ownerClaim0 = $util.defaultIfNull($ctx.identity.claims.get(\\"username\\"), $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:username\\"), \\"___xamznone____\\")) )
    #set( $ownerAllowedFields0 = [] )
    #if( $ownerClaim0 == $ownerEntity0 )
      #if( !$ownerAllowedFields0.isEmpty() )
        $util.qr($allowedFields.addAll($ownerAllowedFields0))
      #else
        #set( $isAuthorized = true )
      #end
    #end
    #if( $util.isNull($ownerEntity0) && !$ctx.args.input.containsKey(\\"owner\\") )
      $util.qr($ctx.args.input.put(\\"owner\\", $ownerClaim0))
      #if( !$ownerAllowedFields0.isEmpty() )
        $util.qr($allowedFields.addAll($ownerAllowedFields0))
      #else
        #set( $isAuthorized = true )
      #end
    #end
  #end
#end
#if( !$isAuthorized && $allowedFields.isEmpty() )
$util.unauthorized()
#end
#if( !$isAuthorized )
  #set( $deniedFields = $util.list.copyAndRemoveAll($inputFields, $allowedFields) )
  #if( $deniedFields.size() > 0 )
    $util.error(\\"Unauthorized on \${deniedFields}\\", \\"Unauthorized\\")
  #end
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization Steps. **"
`;

exports[`amplify add api (GraphQL) init a project and add custom iam roles - local test with gql v2 2`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $inputFields = $util.parseJson($util.toJson($ctx.args.input.keySet())) )
#set( $isAuthorized = false )
#set( $allowedFields = [] )
#if( $util.authType() == \\"IAM Authorization\\" )
  #set( $adminRoles = [\\"myAdminRoleName\\"] )
  #foreach( $adminRole in $adminRoles )
    #if( $ctx.identity.userArn.contains($adminRole) && $ctx.identity.userArn != $ctx.stash.authRole && $ctx.identity.userArn != $ctx.stash.unauthRole )
      #return($util.toJson({}))
    #end
  #end
$util.unauthorized()
#end
#if( $util.authType() == \\"User Pool Authorization\\" )
  #if( !$isAuthorized )
    #set( $ownerEntity0 = $util.defaultIfNull($ctx.args.input.owner, null) )
    #set( $ownerClaim0 = $util.defaultIfNull($ctx.identity.claims.get(\\"username\\"), $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:username\\"), \\"___xamznone____\\")) )
    #set( $ownerAllowedFields0 = [] )
    #if( $ownerClaim0 == $ownerEntity0 )
      #if( !$ownerAllowedFields0.isEmpty() )
        $util.qr($allowedFields.addAll($ownerAllowedFields0))
      #else
        #set( $isAuthorized = true )
      #end
    #end
    #if( $util.isNull($ownerEntity0) && !$ctx.args.input.containsKey(\\"owner\\") )
      $util.qr($ctx.args.input.put(\\"owner\\", $ownerClaim0))
      #if( !$ownerAllowedFields0.isEmpty() )
        $util.qr($allowedFields.addAll($ownerAllowedFields0))
      #else
        #set( $isAuthorized = true )
      #end
    #end
  #end
#end
#if( !$isAuthorized && $allowedFields.isEmpty() )
$util.unauthorized()
#end
#if( !$isAuthorized )
  #set( $deniedFields = $util.list.copyAndRemoveAll($inputFields, $allowedFields) )
  #if( $deniedFields.size() > 0 )
    $util.error(\\"Unauthorized on \${deniedFields}\\", \\"Unauthorized\\")
  #end
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization Steps. **"
`;
